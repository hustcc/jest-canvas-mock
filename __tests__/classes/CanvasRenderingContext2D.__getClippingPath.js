let ctx;
beforeEach(() => {
  // get a new context each test
  ctx = document.createElement('canvas').getContext('2d');
});

afterEach(() => {
  const drawCalls = ctx.__getClippingRegion();
  expect(drawCalls).toMatchSnapshot();
});

describe('__getClippingRegion', () => {
  it('should be empty when there are no path elements', () => {
    ctx.clip();
  });

  it('should store the clipping region', () => {
    ctx.rect(1, 2, 3, 4);
    ctx.arc(1, 2, 3, 4, 5);
    ctx.clip();
  });

  it("shouldn't store the whole clipping region twice when clip is called twice", () => {
    ctx.rect(1, 2, 3, 4);
    ctx.arc(1, 2, 3, 4, 5);
    ctx.clip();
    ctx.rect(1, 2, 3, 4);
    ctx.arc(1, 2, 3, 4, 5);
    ctx.clip();
  });

  it('should save the clipping region correctly when saved', () => {
    ctx.rect(1, 2, 3, 4);
    ctx.arc(1, 2, 3, 4, 5);
    ctx.clip();
    const region = ctx.__getClippingRegion();
    ctx.save();
    expect(region).toStrictEqual(ctx.__getClippingRegion());
  });

  it('should save the clipping region correctly when saved', () => {
    ctx.rect(1, 2, 3, 4);
    ctx.arc(1, 2, 3, 4, 5);
    ctx.clip();
    const region = ctx.__getClippingRegion();
    ctx.save();
    ctx.rect(1, 2, 3, 4);
    ctx.arc(1, 2, 3, 4, 5);
    ctx.clip();
    expect(region).not.toStrictEqual(ctx.__getClippingRegion());
    ctx.restore();
    expect(region).toStrictEqual(ctx.__getClippingRegion());
  });

  it('should delete current clipping region when restored', () => {
    ctx.rect(1, 2, 3, 4);
    ctx.clip();
    ctx.save();
    ctx.rect(1, 2, 3, 4);
    ctx.arc(1, 2, 3, 4, 5);
    ctx.clip();
    ctx.restore();
    ctx.save();
  });
});
