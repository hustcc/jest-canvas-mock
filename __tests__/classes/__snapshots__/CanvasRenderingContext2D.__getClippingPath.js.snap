// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`__getClippingRegion should be empty when there are no path elements 1`] = `Array []`;

exports[`__getClippingRegion should save the clipping region correctly when saved 1`] = `
Array [
  Object {
    "props": Object {
      "height": 4,
      "width": 3,
      "x": 1,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "rect",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 5,
      "radius": 3,
      "startAngle": 4,
      "x": 1,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
]
`;

exports[`__getClippingRegion should save the clipping region correctly when saved 2`] = `
Array [
  Object {
    "props": Object {
      "height": 4,
      "width": 3,
      "x": 1,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "rect",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 5,
      "radius": 3,
      "startAngle": 4,
      "x": 1,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
]
`;

exports[`__getClippingRegion should store the clipping region 1`] = `
Array [
  Object {
    "props": Object {
      "height": 4,
      "width": 3,
      "x": 1,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "rect",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 5,
      "radius": 3,
      "startAngle": 4,
      "x": 1,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
]
`;

exports[`__getClippingRegion shouldn't store the whole clipping region twice when clip is called twice 1`] = `
Array [
  Object {
    "props": Object {
      "height": 4,
      "width": 3,
      "x": 1,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "rect",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 5,
      "radius": 3,
      "startAngle": 4,
      "x": 1,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "height": 4,
            "width": 3,
            "x": 1,
            "y": 2,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "rect",
        },
        Object {
          "props": Object {
            "anticlockwise": false,
            "endAngle": 5,
            "radius": 3,
            "startAngle": 4,
            "x": 1,
            "y": 2,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arc",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clip",
  },
  Object {
    "props": Object {
      "height": 4,
      "width": 3,
      "x": 1,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "rect",
  },
  Object {
    "props": Object {
      "anticlockwise": false,
      "endAngle": 5,
      "radius": 3,
      "startAngle": 4,
      "x": 1,
      "y": 2,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "arc",
  },
]
`;
